### ISSUE TITLE
libstdc++ option for AVR

### ISSUE THREAD

**Issue by Floessie** [crosstool-ng/crosstool-ng#2234]
Hi,

I came across #1569 and `CC_GCC_EXTRA_LIBSTDCXX` and built a GCC 14.2 toolchain for AVR with the recent crosstool-ng master branch and this option set. My expectation was to find `libstdc++` in the toolchain (for `<type_traits>`, `<array>` and such), but to my surprise there wasn't even a trace of `libstdc++` in the install path. I guess, my expectation was wrong, but I'd like to get to know, what `CC_GCC_EXTRA_LIBSTDCXX` is for then.

Thanks,
Flössie

**Comment by QBos07** [comment]
I take another look at this. Things may have changed...

**Comment by QBos07** [comment]
This seems related: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=111639
I'll try using gcc 13.3.0 to see if it work on that older version.

**Comment by QBos07** [comment]
TLDR: Some stuff was lost in an efford to cut compilaiton time. (tbh, it did like x3 the process.) But this wasnt really updated and has a chance to work under certan condition. just that they get lower as time progresses. Essentially i have to scientificly trow stuff at the toolchain to note what does the intendet. For the (hopefully) small time beeing avr is kind of brocken, especially with avr-libc. Expect a fix from me soon (tm) since i take the responsibility here

**Comment by QBos07** [comment]
Got it "working"! (But only in freestanding mode)

1. make a normal config with C++ enabled.
2. tick "Re-enable libstdcxx"
3. add `--enable-libstdcxx --disable-hosted-libstdcxx` to "gcc extra config"
4. do something like this: 
```sh
echo "CT_CC_CORE_NEEDED=y" >> .config
echo "CT_CC_CORE_PASS_2_NEEDED=y" >> .config
```
5. run `ct-ng build`
6. Enjoy your avr libstdcxx toolchain!

You may need to use a high C++ (preferable 20 and up) standart to use all avaible headers in freestanding mode.
Avr is technicly not supported by (especially hosted) libstdc++-v3 right now. But it's getting kind of maintained.
Trying to build a hosted version will get you into multiple problems due to avr-libc not being fully c/posix-compilant.

**Comment by QBos07** [comment]
The PR works!

Got this working:
```c++
#include <type_traits>

int main() {
  return std::alignment_of<int>::value;
}
```

compiled with `avr-g++ -mmcu=atmega32 -std=gnu++20` and tested with simavr+avr-gdb.

**Comment by Floessie** [comment]
Great work, Quentin! :+1: 

With the PR, are the four points for preparation still needed?

In the meantime I found [avr-libstdcpp](https://github.com/modm-io/avr-libstdcpp) which meets my requirements. But it's good to have the option in ct-ng working again. :+1:

**Comment by QBos07** [comment]
The pr has includes all of these points. While the libraries share the same base there different philosophy which is noticeable when time passes on.

**Comment by Floessie** [comment]
@QBos07 `CT_CC_GCC_LIBSTDCXX=y` did the trick. :tada: My fault.

Thanks,
Flössie

**Comment by RRArinc** [comment]
Hi,
In .config file i see CT_CC_GCC_LIBSTDCXX_VERBOSE and not CT_CC_GCC_LIBSTDCXX
Should i set the flag =y or create one flag with name CT_CC_GCC_LIBSTDCXX and set it 'y".
Your input is much appreciated here.

Thanks

**Comment by QBos07** [comment]
Please modify the config only with ct-ng menuconfig if you are not completely sure what you are doing.

The make the libstdcxx point visible you may need to explicitly enable c++ support a bit later down in the gcc options
